metron-platform/metron-parsers/README.md

```diff
@@ -82,6 +82,12 @@ topology in kafka.  Errors are collected with the context of the error
 (e.g. stacktrace) and original message causing the error and sent to an
 `error` queue.  Invalid messages as determined by global validation
 functions are also treated as errors and sent to an `error` queue. 
+
+Multiple sensors can be aggregated into a single Storm topology. When this is done, there will be
+multiple Kafka spouts, but only a single parser bolt which will handle delegating to the correct 
```

Comment:
Correct, it's aggregating existing topics / sensors into a single topology (with multiple spouts).  It's pulling from each individual topic/sensor as a spout, and then passing to a single parser bolt which handles parsing and output as needed.

Reply:
There is another, more likely use case where we have a transport wrapper on another message, and 1 topic split into many parsers as well.  How can we handle that?

Specifically -> Syslog (Many Msg types) -> kafka -> bolt -> Split per message

I expect to add the ability for syslog parsing later, so set that aside.  The issue is we *will* have more than one use case wrt topics.  

I am not going to say this PR needs to address it, but I would want us to understand our path forward and minimize the churn.

It would be best if we did not have to redo this work when accounting for that.

Reply:
@ottobackwards I think that's essentially the parser chaining stuff I added earlier, am I misunderstanding?  The [use-case](https://github.com/apache/metron/tree/master/use-cases/parser_chaining) is using the "we get a ton of types of data in syslog" example.

Reply:
I think I'm misunderstanding between your diagram and this implementation.  There will be one kafka topic monitored by the bolt,  then routed to the right parser, then output to a different spout per parser?

Reply:
Essentially the flow in the syslog case is:

1. Syslog message comes in through the wrapper's topic.
2. The spout for syslog passes it through to the aggregate bolt.
3. The aggregated bolt does its job unwrapping and outputs the inner messages to the appropriate topics for each.
4. The spouts for the inner messages then pick up from the now populated topics.
5. All spouts are now passing messages as they receive them to the aggregate bolt.
6. All messages get delegated and handled as needed.

This implementation doesn't do anything clever around avoiding output topics if the aggregate parser is already contained here.

Reply:
This PR gives us the ability to group the parsers into a single topology if we so desire.  You would still write through to kafka.  So, the topology in the example would have 3 kafka spouts:
* One for monitoring `pix_syslog_router` (the syslog parser aka the routing parser)
* One for monitoring `cisco-5-304`
* One for monitoring `cisco-6-302`

There would be one parser bolt, though, which would handle parsing all 3 sensor types.  That is the contribution of this PR, the ability to determine the parser and filter and field transformations from the input kafka topic and use the appropriate one to parse the messages.  There is not, however, any code here that would bypass the intermediate kafka write (e.g. from the router topology to the individual `cisco-5-304` or `cisco-6-302` topics).  That's a current gap.

Reply:
The bolt is actually executing the parser, not sending it to kafka though.

Let's say that I route all my syslog stuff ( multiple message types ) through 1 kafka topic ( not tied to _ANY_ sensor.  I would want to point this bolt at that one topic, and have it route to multiple parsers. I think then they all just go to enrichment?

Reply:
Yeah, in order to do that, we'd need to execute the DAG without the intermediate kafka step.  That'd be a follow-on.  This is the first step in that.  We have all the information here (the input -> output mapping of kafka topics), so it shouldn't be so hard to intervene and cut the intermediate kafka writing step out.

Reply:
You also might want 2 different parsers that source from the same kafka topic (think: 1 parser to send the data to enrichment and 1 parser to send to hbase as a streaming enrichment for authentication data)

Reply:
@justinleet can you maybe create a data flow diagram or sequence diagram that shows a syslog record from the use-case flowing through this topology and add it to the use-case around parser chaining?

It'd be something like, given a `cisco-6-302` record, it'll go:
* From NiFi to the `pix_syslog_router` kafka topic
* From the `pix_syslog_router` kafka topic to the `pix_syslog_router` spout in the aggregated storm topology
* From the `pix_syslog_router` kafka spout to the parser bolt, which will run the `pix_syslog_router` Grok parser and write out to the `cisco-6-302` kafka topic
* From the `cisco-6-302` kafka topic to the `cisco-6-302` spout in the aggregated storm topology
* From the `cisco-6-302` kafka spout to the `cisco-6-302` Grok parser and write out to the `enrichments` kafka topic, where it's picked up by the enrichment topology.

Eventually, we should consider taking out the writing to the `cisco-6-302` topic (optionally), but even eventually there may be value in those intermediate kafka topics due to how users may want to group sensors (e.g. grouping may be done via velocity or scalability requirements, rather than logical connection).

Reply:
Yeah, I'll go ahead and add a diagram to the doc, and flesh out that explanation

Reply:
Added a diagram [here](https://github.com/justinleet/metron/blob/995fdd7d4f8ce3b49c7e29aaa7842eb99137a813/use-cases/parser_chaining/README.md#aggregated-parsers-with-parser-chaining), and also linked to it in the parsers readme.